File: java/lib/src/main/java/io/openweb3/pay/Invoice.java
package io.openweb3.pay;

import io.openweb3.pay.exceptions.ApiException;
import io.openweb3.pay.internal.api.InvoiceApi;
import io.openweb3.pay.models.InvoiceIn;
import io.openweb3.pay.models.InvoiceOut;
import io.openweb3.pay.models.ListResponseInvoiceOut;

public final class Invoice {
	private final InvoiceApi api;

	Invoice() {
		api = new InvoiceApi();
	}

	public ListResponseInvoiceOut list(final String appId, final InvoiceListOptions options) throws ApiException {
		try {
			return api.v1InvoiceList(appId, options.getLimit(), options.getOffset(), options.getUserId(), options.getChannel());
		} catch (io.openweb3.pay.internal.ApiException e) {
			throw Utils.WrapInternalApiException(e);
		}
	}

	public InvoiceOut create(final String appId, final InvoiceIn InvoiceIn) throws ApiException {
		return this.create(appId, InvoiceIn, new PostOptions());
	}

	public InvoiceOut create(final String appId, final InvoiceIn InvoiceIn, final PostOptions options) throws ApiException {
		try {
			return api.v1InvoiceCreate(appId, InvoiceIn);
		} catch (io.openweb3.pay.internal.ApiException e) {
			throw Utils.WrapInternalApiException(e);
		}
	}

	public InvoiceOut getOrCreate(final String appId, final InvoiceIn InvoiceIn) throws ApiException {
		return this.getOrCreate(appId, InvoiceIn, new PostOptions());
	}

	public InvoiceOut getOrCreate(final String appId, final InvoiceIn InvoiceIn, final PostOptions options) throws ApiException {
		try {
			return api.v1InvoiceCreate(appId, InvoiceIn);
		} catch (io.openweb3.pay.internal.ApiException e) {
			throw Utils.WrapInternalApiException(e);
		}
	}

	// retrieve
	public InvoiceOut get(final String appId, final String idOrUid) throws ApiException {
		try {
			return api.v1InvoiceGet(appId, idOrUid);
		} catch (io.openweb3.pay.internal.ApiException e) {
			throw Utils.WrapInternalApiException(e);
		}
	}
}

File: go/invoice.go
package pay

import (
	"context"

	"github.com/openweb3-io/pay-openapi/go/internal/openapi"
)

type (
	ListResponseInvoiceOut = openapi.PageInvoice
	InvoiceIn              = openapi.CreateInvoiceRequest
	InvoiceOut             = openapi.Invoice
)

type Invoice struct {
	api *openapi.APIClient
}

type InvoiceListOptions struct {
	Offset *int32
	Limit  *int32
	UserId *string
}

func (e *Invoice) List(ctx context.Context, appId string, options *InvoiceListOptions) (*ListResponseInvoiceOut, error) {
	req := e.api.InvoiceApi.V1InvoicesList(ctx, appId)
	if options != nil {
		if options.Limit != nil {
			req = req.Limit(*options.Limit)
		}
		if options.Offset != nil {
			req = req.Offset(*options.Offset)
		}
	}
	out, res, err := req.Execute()
	if err != nil {
		return nil, wrapError(err, res)
	}
	ret := ListResponseInvoiceOut(out)
	return &ret, nil
}

func (e *Invoice) Create(ctx context.Context, appId string, invoiceIn *InvoiceIn) (*InvoiceOut, error) {
	return e.CreateWithOptions(ctx, appId, invoiceIn, nil)
}

func (e *Invoice) CreateWithOptions(ctx context.Context, appId string, invoiceIn *InvoiceIn, options *PostOptions) (*InvoiceOut, error) {
	req := e.api.InvoiceApi.V1InvoicesCreate(ctx, appId)
	req = req.CreateInvoiceRequest(*invoiceIn)

	out, res, err := req.Execute()
	if err != nil {
		return nil, wrapError(err, res)
	}
	ret := InvoiceOut(out)
	return &ret, nil
}

func (e *Invoice) Get(ctx context.Context, appId string, invoiceId string) (*InvoiceOut, error) {
	req := e.api.InvoiceApi.V1InvoicesRetrieve(ctx, appId, invoiceId)

	out, res, err := req.Execute()
	if err != nil {
		return nil, wrapError(err, res)
	}
	ret := InvoiceOut(out)

	return &ret, nil
}
